### ISSUE TITLE
Resource Variable availability undefined

### ISSUE THREAD

**Issue by thisjustin816** [MicrosoftDocs/azure-devops-yaml-schema#29]
It's not defined when resource variables are available. For example, none of `$[ variables.resources.pipeline.ResourceAlias.runName ]`,  `$[ variables['resources.pipeline.ResourceAlias.runName']]`, or `${{ variables.resources.pipeline.ResourceAlias.runName }}` work when trying to map it to a YAML variable block.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: bb60190b-fc26-9eb2-70c3-a8b1a0227658
* Version Independent ID: bb60190b-fc26-9eb2-70c3-a8b1a0227658
* Content: [resources.pipelines.pipeline definition](https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema/resources-pipelines-pipeline?view=azure-pipelines#the-pipeline-resource-metadata-as-predefined-variables)
* Content Source: [content/resources-pipelines-pipeline.md](https://github.com/MicrosoftDocs/azure-devops-yaml-schema/blob/main/content/resources-pipelines-pipeline.md)
* Product: **devops**
* Technology: **devops-cicd**
* GitHub Login: @steved0x
* Microsoft Alias: **sdanie**

**Comment by steved0x** [comment]
@thisjustin816 Let me see if I can get some more detail on that.

**Comment by steved0x** [comment]
@thisjustin816 Sorry for the long delay in responding to this issue. Were you able to get this working?

These variables should be available in the jobs of a pipeline (as the job is running) that was triggered by a pipeline that has one or more pipeline resources. For example, I have the following test pipeline:

```
resources:
 pipelines:
   - pipeline: source-pipeline
     source: PipelineTriggerSource
     project: FabrikamFiber
     trigger: true
   - pipeline: other-project-pipeline
     source: PipelineTriggerFromOtherProject
     project: FabrikamRepo
     trigger: true

trigger: none # Only trigger with pipeline resource trigger

pool:
  vmImage: ubuntu-latest

steps:
- script: echo Hello, world!
  displayName: 'Run a one-line script'

- bash: echo $(resources.pipeline.source-pipeline.projectName)
- bash: printenv | sort
```

The first bash task outputs the value of the projectName variable of the pipeline resource named `source-pipeline`.

The second bash task outputs all of the environment variables (which includes pipeline variables). This pipeline has two pipeline resources, and in the listing of all environment variables it includes these two:

```
RESOURCES_PIPELINE_OTHER-PROJECT-PIPELINE_PROJECTNAME=FabrikamRepo
RESOURCES_PIPELINE_SOURCE-PIPELINE_PROJECTNAME=FabrikamFiber
```

Note the naming change when they are listed as environment variables, the `.` is replaced with `_` and the letters are all capitalized.

I am adding this example to the article and including the steps for outputting all the variables, and when that article update goes live, it will automatically close this GH issue. Please open a new issue if this info doesn't resolve the question for you.

**Comment by thisjustin816** [comment]
> @thisjustin816 Sorry for the long delay in responding to this issue. Were you able to get this working?
> 
> These variables should be available in the jobs of a pipeline (as the job is running) that was triggered by a pipeline that has one or more pipeline resources. For example, I have the following test pipeline:
> 
> ```
> resources:
>  pipelines:
>    - pipeline: source-pipeline
>      source: PipelineTriggerSource
>      project: FabrikamFiber
>      trigger: true
>    - pipeline: other-project-pipeline
>      source: PipelineTriggerFromOtherProject
>      project: FabrikamRepo
>      trigger: true
> 
> trigger: none # Only trigger with pipeline resource trigger
> 
> pool:
>   vmImage: ubuntu-latest
> 
> steps:
> - script: echo Hello, world!
>   displayName: 'Run a one-line script'
> 
> - bash: echo $(resources.pipeline.source-pipeline.projectName)
> - bash: printenv | sort
> ```
> 
> The first bash task outputs the value of the projectName variable of the pipeline resource named `source-pipeline`.
> 
> The second bash task outputs all of the environment variables (which includes pipeline variables). This pipeline has two pipeline resources, and in the listing of all environment variables it includes these two:
> 
> ```
> RESOURCES_PIPELINE_OTHER-PROJECT-PIPELINE_PROJECTNAME=FabrikamRepo
> RESOURCES_PIPELINE_SOURCE-PIPELINE_PROJECTNAME=FabrikamFiber
> ```
> 
> Note the naming change when they are listed as environment variables, the `.` is replaced with `_` and the letters are all capitalized.
> 
> I am adding this example to the article and including the steps for outputting all the variables, and when that article update goes live, it will automatically close this GH issue. Please open a new issue if this info doesn't resolve the question for you.

This doesn't address the issue. I opened this for some more clarity when using those variables outside of a task. It's not clear where the valid use cases are, or if you can use them at all.

**Comment by steved0x** [comment]
@thisjustin816 Thank you for the additional context. I will find out and report back.

**Comment by steved0x** [comment]
> I am adding this example to the article and including the steps for outputting all the variables, and when that article update goes live, it will automatically close this GH issue. Please open a new issue if this info doesn't resolve the question for you.

I am adding the example as I have had some other questions about the naming of the different variables, but I have disabled the "auto-closing" of this issue, until I am able to get more information from the product team about your scenario.

**Comment by steved0x** [comment]
@thisjustin816 I have an inquiry to the dev team to see if there is anything "special" about the pipeline resource variables, but while I wait for that info, I can provide the "standard" info about using variables in pipelines.

The syntax depends on how you want to use it, for these, it would be either macro syntax `$(var)` to get processed at rutime, or runtime syntax `$[variables.var]` to be processed at runtime when used in conditions or expressions. (but not template syntax `${{ variables.var }}` which is processed at compile time before the pipeline is triggered)

(These articles have some more info and I will link them from the pipeline resources article)
https://learn.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#understand-variable-syntax
https://learn.microsoft.com/en-us/azure/devops/pipelines/process/conditions?view=azure-devops&tabs=yaml%2Cstages
https://learn.microsoft.com/en-us/azure/devops/pipelines/process/expressions?view=azure-devops

So a contrived example of a condition would be like this, using the runtime syntax.

```
condition: eq(variables['resources.pipeline.source-pipeline.projectName'], 'FabrikamFiber')
```

The whole pipeline is like this
```
resources:
 pipelines:
   - pipeline: source-pipeline
     source: PipelineTriggerSource
     project: FabrikamFiber
     trigger: true
   - pipeline: other-project-pipeline
     source: PipelineTriggerFromOtherProject
     project: FabrikamRepo
     trigger: true

trigger: none # Only trigger with pipeline resource trigger

pool:
  vmImage: windows-latest

jobs:
- job: Test1
  # Only run this job if the expression is true
  condition: eq(variables['resources.pipeline.source-pipeline.projectName'], 'FabrikamFiber')
  steps:
  - bash: echo $(resources.pipeline.source-pipeline.projectName)
  - bash: printenv | sort
```

The condition is always true so it runs (it is a very contrived example to show syntax only)

> So can they be used in condition statements, build and environment names

Yes for condition statements, and other values in the pipeline (like if you wanted to use an expression containing one of these variables for a pipeline run number.

But they wouldn't be able to be used to resolve environment names, because resources are authorized before a stage can start, and so stage and job level variables aren't available. Pipeline-level variables can be used, but only those variables explicitly included in the pipeline. (I am uclear about this part - are these considered pipeline level variables- and I am waiting on clarification)
https://learn.microsoft.com/en-us/azure/devops/pipelines/process/runs?view=azure-devops#process-the-pipeline

Do you have a specific scenario where you are wanting to use these variables? I can share that with the thread I have started with the team.

Thank you,
Steve

**Comment by steved0x** [comment]
Thank you for your feedback. We are migrating this issue to our new documentation feedback system, and closing this issue.
 
As we phase out GitHub issues for this repository, we encourage you to propose changes using public PRs or submit feedback through our updated control at the top of the page. For more information on our feedback systems, see https://aka.ms/contentuserfeedback.
